pkg/service/server.go

```diff
@@ -427,51 +446,87 @@ func (mgr *Manager) drainLoadbalancerTarget(event *LifecycleEvent) error {
 		activeLoadBalancers = append(activeLoadBalancers, elbName)
 	}
 
-	workQueueLength := len(activeTargetGroups) + len(activeLoadBalancers)
-
 	// create goroutine per target group with target match
+	workQueueLength := len(activeTargetGroups) + len(activeLoadBalancers)
 	wg.Add(workQueueLength)
+	errChannel := make(chan error, workQueueLength*2)
 
-	// sleep for random jitter per goroutine
-	waitJitter(ThreadJitterRangeSeconds)
+	log.Infof("found %v target groups & %v classic-elb for instance %v", len(activeTargetGroups), len(elbDescriptions), instanceID)
 
-	// handle classic load balancers
-	for _, elbName := range activeLoadBalancers {
+	log.Infof("starting deregistration for %v", instanceID)
+	// handle classic load balancers deregistration
+	deregisteredLoadBalancers := []string{}
+	for i, elbName := range activeLoadBalancers {
 
-		// sleep for random jitter between iterations of deregister
-		waitJitter(DeregisterJitterRangeSeconds)
+		if event.eventCompleted {
+			return errors.New("event finished execution during deregistration")
+		}
 
-		go func(elbName, instance string) {
-			defer wg.Done()
-			// deregister from classic-elb
-			log.Debugf("found %v in classic-elb %v, will deregister", instance, elbName)
-			err = deregisterInstance(elbClient, elbName, instance)
-			if err != nil {
-				errChannel <- err
-				msg := fmt.Sprintf(EventMessageInstanceDeregisterFailed, instance, elbName, err)
-				msgFields := map[string]string{
-					"elbName":       elbName,
-					"ec2InstanceId": instance,
-					"elbType":       "classic-elb",
-					"details":       msg,
-				}
-				publishKubernetesEvent(kubeClient, newKubernetesEvent(EventReasonInstanceDeregisterFailed, msgFields, event.referencedNode.Name))
-				return
+		// sleep for random jitter per iteration
+		waitJitter(IterationJitterRangeSeconds)
+		log.Debugf("deregistering %v from %v (%v/%v)", instanceID, elbName, i+1, len(activeTargetGroups))
+		err = deregisterInstance(elbClient, elbName, instanceID)
+		if err != nil {
+			log.Info("deregister failed")
+			errChannel <- err
+			log.Info("after")
+			msg := fmt.Sprintf(EventMessageInstanceDeregisterFailed, instanceID, elbName, err)
+			msgFields := map[string]string{
+				"elbName":       elbName,
+				"ec2InstanceId": instanceID,
+				"elbType":       "classic-elb",
+				"details":       msg,
 			}
+			publishKubernetesEvent(kubeClient, newKubernetesEvent(EventReasonInstanceDeregisterFailed, msgFields, event.referencedNode.Name))
+			continue
+		}
+		deregisteredLoadBalancers = append(deregisteredLoadBalancers, elbName)
+	}
+
+	// handle v2 load balancers deregistration
+	deregisteredTargetGroups := map[string]int64{}
+	currentDeregistering := 0
+	for arn, port := range activeTargetGroups {
+
+		if event.eventCompleted {
+			return errors.New("event finished execution during deregistration")
+		}
+
+		currentDeregistering++
+		// sleep for random jitter per iteration
+		waitJitter(IterationJitterRangeSeconds)
```

Comment:
anyway to move this mostly duplicated (with lines 466-483) code into a shared location and called with params as necessary?

Reply:
A bigger refactor is needed to move this code around, let's do this down the road.